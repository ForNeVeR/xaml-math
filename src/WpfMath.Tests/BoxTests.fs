module WpfMath.Tests.BoxTests

open System

open FSharp.Core.Fluent
open Xunit

open WpfMath
open WpfMath.Atoms
open WpfMath.Boxes
open WpfMath.Tests.ApprovalTestUtils

let private parse(text: string) =
    let parser = TexFormulaParser()
    let result = parser.Parse text
    result.RootAtom

let private src (string: string) (start: int) (len: int) = SourceSpan("User input", string, start, len)

let private environment =
    let mathFont = DefaultTexFont 20.0
    let textFont = TexFormula.GetSystemFont("Arial", 20.0)
    TexEnvironment(TexStyle.Display, mathFont, textFont)

[<Fact>]
let ``AccentedAtom should have a skew according to the char``() =
    let topAtom = parse @"\bar{\bar{x}}" :?> AccentedAtom
    let childAtom = topAtom.BaseAtom :?> AccentedAtom

    let topBox = topAtom.CreateBox(environment).Children.[0]
    let childBox = childAtom.CreateBox(environment).Children.[0]

    Assert.Equal(topBox.Shift, childBox.Shift)

[<Fact>]
let ``Box for \text{æ,} should be created successfully``() =
    let atom = parse @"\text{æ,}"
    let box = atom.CreateBox(environment)
    Assert.NotNull(box)

[<Fact>]
let ``ScriptsAtom should set Shift on the created box when creating box without any sub- or superscript``() =
    Utils.initializeFontResourceLoading()

    let baseAtom = CharAtom(src "x" 0 1, 'x')
    let scriptsAtom = ScriptsAtom(null, baseAtom, null, null)

    let box = scriptsAtom.CreateBox(environment)

    let expectedShift = -(box.Height + box.Depth) / 2.0 - environment.MathFont.GetAxisHeight(environment.Style)
    Assert.Equal(expectedShift, box.Shift)

[<Fact>]
let ``RowAtom creates boxes with proper sources``() =
    let source = "2+2"
    let src = src source
    let parser = TexFormulaParser()
    let formula = parser.Parse source
    let box = formula.CreateBox environment :?> HorizontalBox
    let chars = box.Children.filter (fun x -> x :? CharBox)
    Assert.Collection (
        chars,
        Action<_>(fun (x: Box) -> Assert.Equal(src 0 1, x.Source)),
        Action<_>(fun (x: Box) -> Assert.Equal(src 1 1, x.Source)),
        Action<_>(fun (x: Box) -> Assert.Equal(src 2 1, x.Source)))

[<Fact>]
let ``BigOperatorAtom creates a box with proper sources``() =
    let source = @"\int_a^b"
    let src = src source
    let parser = TexFormulaParser()
    let formula = parser.Parse source
    let box = formula.CreateBox environment :?> VerticalBox

    let charBoxes =
        box.Children
            .filter(fun x -> x :? HorizontalBox)
            .collect(fun x -> x.Children.filter (fun y -> y :? CharBox))
            .toList()

    Assert.Collection (
        charBoxes,
        Action<_>(fun (x: Box) -> Assert.Equal(src 7 1, x.Source)),
        Action<_>(fun (x: Box) -> Assert.Equal(src 1 3, x.Source)),
        Action<_>(fun (x: Box) -> Assert.Equal(src 5 1, x.Source)))

[<Fact>]
let ``Cyrillic followed by Latin should be rendered properly``() =
    Utils.initializeFontResourceLoading()
    let atom = parse @"\text{Ц}V"
    let box = atom.CreateBox environment
    Assert.NotNull(box)

let private verifyBox source =
    let atom = parse source
    let box = atom.CreateBox environment
    verifyObject box

[<Fact>]
let simpleMatrixBox() =
    verifyBox @"\pmatrix{2 & 2 \\ 2 & 2}"

[<Fact>]
let casesBox() =
    verifyBox @"\cases{a \\ b \\ c}"

[<Fact>]
let nestedMatrixBox() =
    verifyBox @"\matrix{ 1 & 2 & 3 \\ 4 & {\matrix{ 5 \\ 6 }} & 7 }"

[<Fact>]
let wideItemInMatrixBox() =
    verifyBox @"x = \pmatrix{0 & -r & 0 \\ 0 & 0 & -r sin^2(\theta)}"

[<Fact>]
let shortСommandForThinspace(): unit =
    verifyBox @"\,"

[<Fact>]
let shortСommandForNotEqual(): unit =
    verifyBox @"\neq"

[<Fact>]
let emptyColorbox(): unit =
    verifyBox @"\colorbox{red}{}"

[<Fact>]
let emptyMathrm(): unit =
    verifyBox @"\mathrm{}"

[<Fact>]
let emptyCommandText(): unit =
    verifyBox @"\text{}"

[<Fact>]
let emptyColorRed(): unit =
    verifyBox @"\color{red}{}"
