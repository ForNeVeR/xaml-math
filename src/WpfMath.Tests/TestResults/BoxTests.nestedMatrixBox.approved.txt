{
  "Children": [
    {
      "Children": [
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.175,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.175,
              "Depth": 0.0,
              "Shift": 0.35
            },
            {
              "Character": {
                "Character": "1",
                "Font": "pack://application:,,,/WpfMath;component/Fonts/cmr10.ttf",
                "Size": 1.0,
                "Metrics": {
                  "Width": 0.500002,
                  "Height": 0.644444,
                  "Depth": 0.0,
                  "Italic": 0.0
                },
                "FontId": 1
              },
              "Children": [],
              "Source": {
                "Start": 9,
                "End": 10,
                "Length": 1,
                "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
                "SourceName": "User input"
              },
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.644444,
              "TotalWidth": 0.500002,
              "Italic": 0.0,
              "Width": 0.500002,
              "Height": 0.644444,
              "Depth": 0.0,
              "Shift": 0.0
            },
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.175,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.175,
              "Depth": 0.0,
              "Shift": 0.35
            }
          ],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.994444,
          "TotalWidth": 0.500002,
          "Italic": 0.0,
          "Width": 0.500002,
          "Height": 0.994444,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.35,
          "Italic": 0.0,
          "Width": 0.35,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.175,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.175,
              "Depth": 0.0,
              "Shift": 0.35
            },
            {
              "Character": {
                "Character": "2",
                "Font": "pack://application:,,,/WpfMath;component/Fonts/cmr10.ttf",
                "Size": 1.0,
                "Metrics": {
                  "Width": 0.500002,
                  "Height": 0.644444,
                  "Depth": 0.0,
                  "Italic": 0.0
                },
                "FontId": 1
              },
              "Children": [],
              "Source": {
                "Start": 13,
                "End": 14,
                "Length": 1,
                "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
                "SourceName": "User input"
              },
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.644444,
              "TotalWidth": 0.500002,
              "Italic": 0.0,
              "Width": 0.500002,
              "Height": 0.644444,
              "Depth": 0.0,
              "Shift": 0.0
            },
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.175,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.175,
              "Depth": 0.0,
              "Shift": 0.35
            }
          ],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.994444,
          "TotalWidth": 0.500002,
          "Italic": 0.0,
          "Width": 0.500002,
          "Height": 0.994444,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.35,
          "Italic": 0.0,
          "Width": 0.35,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.175,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.175,
              "Depth": 0.0,
              "Shift": 0.35
            },
            {
              "Character": {
                "Character": "3",
                "Font": "pack://application:,,,/WpfMath;component/Fonts/cmr10.ttf",
                "Size": 1.0,
                "Metrics": {
                  "Width": 0.500002,
                  "Height": 0.644444,
                  "Depth": 0.0,
                  "Italic": 0.0
                },
                "FontId": 1
              },
              "Children": [],
              "Source": {
                "Start": 17,
                "End": 18,
                "Length": 1,
                "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
                "SourceName": "User input"
              },
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.644444,
              "TotalWidth": 0.500002,
              "Italic": 0.0,
              "Width": 0.500002,
              "Height": 0.644444,
              "Depth": 0.0,
              "Shift": 0.0
            },
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.175,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.175,
              "Depth": 0.0,
              "Shift": 0.35
            }
          ],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.994444,
          "TotalWidth": 0.500002,
          "Italic": 0.0,
          "Width": 0.500002,
          "Height": 0.994444,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        }
      ],
      "Source": null,
      "Foreground": null,
      "Background": null,
      "TotalHeight": 0.994444,
      "TotalWidth": 2.900006,
      "Italic": 0.0,
      "Width": 2.900006,
      "Height": 0.994444,
      "Depth": 0.0,
      "Shift": 0.0
    },
    {
      "Children": [
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.847222,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.847222,
              "Depth": 0.0,
              "Shift": 0.35
            },
            {
              "Character": {
                "Character": "4",
                "Font": "pack://application:,,,/WpfMath;component/Fonts/cmr10.ttf",
                "Size": 1.0,
                "Metrics": {
                  "Width": 0.500002,
                  "Height": 0.644444,
                  "Depth": 0.0,
                  "Italic": 0.0
                },
                "FontId": 1
              },
              "Children": [],
              "Source": {
                "Start": 22,
                "End": 23,
                "Length": 1,
                "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
                "SourceName": "User input"
              },
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.644444,
              "TotalWidth": 0.500002,
              "Italic": 0.0,
              "Width": 0.500002,
              "Height": 0.644444,
              "Depth": 0.0,
              "Shift": 0.0
            },
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.847222,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.847222,
              "Depth": 0.0,
              "Shift": 0.35
            }
          ],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 2.338888,
          "TotalWidth": 0.500002,
          "Italic": 0.0,
          "Width": 0.500002,
          "Height": 2.338888,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.175,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.175,
              "Depth": 0.0,
              "Shift": 0.35
            },
            {
              "Children": [
                {
                  "Children": [
                    {
                      "Children": [],
                      "Source": null,
                      "Foreground": null,
                      "Background": null,
                      "TotalHeight": 0.0,
                      "TotalWidth": 0.175,
                      "Italic": 0.0,
                      "Width": 0.175,
                      "Height": 0.0,
                      "Depth": 0.0,
                      "Shift": 0.0
                    },
                    {
                      "Children": [
                        {
                          "Children": [],
                          "Source": null,
                          "Foreground": null,
                          "Background": null,
                          "TotalHeight": 0.175,
                          "TotalWidth": 0.0,
                          "Italic": 0.0,
                          "Width": 0.0,
                          "Height": 0.175,
                          "Depth": 0.0,
                          "Shift": 0.35
                        },
                        {
                          "Character": {
                            "Character": "5",
                            "Font": "pack://application:,,,/WpfMath;component/Fonts/cmr10.ttf",
                            "Size": 1.0,
                            "Metrics": {
                              "Width": 0.500002,
                              "Height": 0.644444,
                              "Depth": 0.0,
                              "Italic": 0.0
                            },
                            "FontId": 1
                          },
                          "Children": [],
                          "Source": {
                            "Start": 36,
                            "End": 37,
                            "Length": 1,
                            "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
                            "SourceName": "User input"
                          },
                          "Foreground": null,
                          "Background": null,
                          "TotalHeight": 0.644444,
                          "TotalWidth": 0.500002,
                          "Italic": 0.0,
                          "Width": 0.500002,
                          "Height": 0.644444,
                          "Depth": 0.0,
                          "Shift": 0.0
                        },
                        {
                          "Children": [],
                          "Source": null,
                          "Foreground": null,
                          "Background": null,
                          "TotalHeight": 0.175,
                          "TotalWidth": 0.0,
                          "Italic": 0.0,
                          "Width": 0.0,
                          "Height": 0.175,
                          "Depth": 0.0,
                          "Shift": 0.35
                        }
                      ],
                      "Source": null,
                      "Foreground": null,
                      "Background": null,
                      "TotalHeight": 0.994444,
                      "TotalWidth": 0.500002,
                      "Italic": 0.0,
                      "Width": 0.500002,
                      "Height": 0.994444,
                      "Depth": 0.0,
                      "Shift": 0.0
                    },
                    {
                      "Children": [],
                      "Source": null,
                      "Foreground": null,
                      "Background": null,
                      "TotalHeight": 0.0,
                      "TotalWidth": 0.175,
                      "Italic": 0.0,
                      "Width": 0.175,
                      "Height": 0.0,
                      "Depth": 0.0,
                      "Shift": 0.0
                    }
                  ],
                  "Source": null,
                  "Foreground": null,
                  "Background": null,
                  "TotalHeight": 0.994444,
                  "TotalWidth": 0.850002,
                  "Italic": 0.0,
                  "Width": 0.850002,
                  "Height": 0.994444,
                  "Depth": 0.0,
                  "Shift": 0.0
                },
                {
                  "Children": [
                    {
                      "Children": [],
                      "Source": null,
                      "Foreground": null,
                      "Background": null,
                      "TotalHeight": 0.0,
                      "TotalWidth": 0.175,
                      "Italic": 0.0,
                      "Width": 0.175,
                      "Height": 0.0,
                      "Depth": 0.0,
                      "Shift": 0.0
                    },
                    {
                      "Children": [
                        {
                          "Children": [],
                          "Source": null,
                          "Foreground": null,
                          "Background": null,
                          "TotalHeight": 0.175,
                          "TotalWidth": 0.0,
                          "Italic": 0.0,
                          "Width": 0.0,
                          "Height": 0.175,
                          "Depth": 0.0,
                          "Shift": 0.35
                        },
                        {
                          "Character": {
                            "Character": "6",
                            "Font": "pack://application:,,,/WpfMath;component/Fonts/cmr10.ttf",
                            "Size": 1.0,
                            "Metrics": {
                              "Width": 0.500002,
                              "Height": 0.644444,
                              "Depth": 0.0,
                              "Italic": 0.0
                            },
                            "FontId": 1
                          },
                          "Children": [],
                          "Source": {
                            "Start": 41,
                            "End": 42,
                            "Length": 1,
                            "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
                            "SourceName": "User input"
                          },
                          "Foreground": null,
                          "Background": null,
                          "TotalHeight": 0.644444,
                          "TotalWidth": 0.500002,
                          "Italic": 0.0,
                          "Width": 0.500002,
                          "Height": 0.644444,
                          "Depth": 0.0,
                          "Shift": 0.0
                        },
                        {
                          "Children": [],
                          "Source": null,
                          "Foreground": null,
                          "Background": null,
                          "TotalHeight": 0.175,
                          "TotalWidth": 0.0,
                          "Italic": 0.0,
                          "Width": 0.0,
                          "Height": 0.175,
                          "Depth": 0.0,
                          "Shift": 0.35
                        }
                      ],
                      "Source": null,
                      "Foreground": null,
                      "Background": null,
                      "TotalHeight": 0.994444,
                      "TotalWidth": 0.500002,
                      "Italic": 0.0,
                      "Width": 0.500002,
                      "Height": 0.994444,
                      "Depth": 0.0,
                      "Shift": 0.0
                    },
                    {
                      "Children": [],
                      "Source": null,
                      "Foreground": null,
                      "Background": null,
                      "TotalHeight": 0.0,
                      "TotalWidth": 0.175,
                      "Italic": 0.0,
                      "Width": 0.175,
                      "Height": 0.0,
                      "Depth": 0.0,
                      "Shift": 0.0
                    }
                  ],
                  "Source": null,
                  "Foreground": null,
                  "Background": null,
                  "TotalHeight": 0.994444,
                  "TotalWidth": 0.850002,
                  "Italic": 0.0,
                  "Width": 0.850002,
                  "Height": 0.994444,
                  "Depth": 0.0,
                  "Shift": 0.0
                }
              ],
              "Source": {
                "Start": 28,
                "End": 44,
                "Length": 16,
                "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
                "SourceName": "User input"
              },
              "Foreground": null,
              "Background": null,
              "TotalHeight": 1.988888,
              "TotalWidth": 0.850002,
              "Italic": 0.0,
              "Width": 0.850002,
              "Height": 1.244444,
              "Depth": 0.744444,
              "Shift": 0.0
            },
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.175,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.175,
              "Depth": 0.0,
              "Shift": 0.35
            }
          ],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 2.338888,
          "TotalWidth": 0.850002,
          "Italic": 0.0,
          "Width": 0.850002,
          "Height": 2.338888,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.847222,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.847222,
              "Depth": 0.0,
              "Shift": 0.35
            },
            {
              "Character": {
                "Character": "7",
                "Font": "pack://application:,,,/WpfMath;component/Fonts/cmr10.ttf",
                "Size": 1.0,
                "Metrics": {
                  "Width": 0.500002,
                  "Height": 0.644444,
                  "Depth": 0.0,
                  "Italic": 0.0
                },
                "FontId": 1
              },
              "Children": [],
              "Source": {
                "Start": 48,
                "End": 49,
                "Length": 1,
                "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
                "SourceName": "User input"
              },
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.644444,
              "TotalWidth": 0.500002,
              "Italic": 0.0,
              "Width": 0.500002,
              "Height": 0.644444,
              "Depth": 0.0,
              "Shift": 0.0
            },
            {
              "Children": [],
              "Source": null,
              "Foreground": null,
              "Background": null,
              "TotalHeight": 0.847222,
              "TotalWidth": 0.0,
              "Italic": 0.0,
              "Width": 0.0,
              "Height": 0.847222,
              "Depth": 0.0,
              "Shift": 0.35
            }
          ],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 2.338888,
          "TotalWidth": 0.500002,
          "Italic": 0.0,
          "Width": 0.500002,
          "Height": 2.338888,
          "Depth": 0.0,
          "Shift": 0.0
        },
        {
          "Children": [],
          "Source": null,
          "Foreground": null,
          "Background": null,
          "TotalHeight": 0.0,
          "TotalWidth": 0.175,
          "Italic": 0.0,
          "Width": 0.175,
          "Height": 0.0,
          "Depth": 0.0,
          "Shift": 0.0
        }
      ],
      "Source": null,
      "Foreground": null,
      "Background": null,
      "TotalHeight": 2.338888,
      "TotalWidth": 2.900006,
      "Italic": 0.0,
      "Width": 2.900006,
      "Height": 2.338888,
      "Depth": 0.0,
      "Shift": 0.0
    }
  ],
  "Source": {
    "Start": 1,
    "End": 51,
    "Length": 50,
    "Source": "\\matrix{ 1 & 2 & 3 \\\\ 4 & {\\matrix{ 5 \\\\ 6 }} & 7 }",
    "SourceName": "User input"
  },
  "Foreground": null,
  "Background": null,
  "TotalHeight": 3.333332,
  "TotalWidth": 2.900006,
  "Italic": 0.0,
  "Width": 2.900006,
  "Height": 1.916666,
  "Depth": 1.416666,
  "Shift": 0.0
}